diff -Naur Browse-157.orig/downloadmanager.py Browse-157/downloadmanager.py
--- Browse-157.orig/downloadmanager.py	2014-10-02 14:39:49.000000000 +0200
+++ Browse-157/downloadmanager.py	2014-11-25 15:19:45.259127938 +0100
@@ -203,8 +203,8 @@
 
             self.dl_jobject.metadata['title'] = \
                 self._download.get_suggested_filename()
-            self.dl_jobject.metadata['description'] = _('From: %s') \
-                % self._source
+            self.dl_jobject.metadata['description'] = _('From: %(source)s') \
+                % {'source': self._source}
             self.dl_jobject.metadata['progress'] = '100'
             self.dl_jobject.file_path = self._dest_path
 
