From 8d13efe154b0bfe4b14a39999a0bb9b0db292e20 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Gy=C3=B6rgy=20Ball=C3=B3?= <ballogyor@gmail.com>
Date: Wed, 23 Mar 2016 01:47:38 +0100
Subject: [PATCH] Add missing parameter to Hungarian translation

---
 po/hu.po | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/po/hu.po b/po/hu.po
index aca808d..27c277b 100644
--- a/po/hu.po
+++ b/po/hu.po
@@ -93,7 +93,7 @@ msgstr "Mutatás a naplóban"
 #: downloadmanager.py:184 pdfviewer.py:589
 #, python-format
 msgid "From: %s"
-msgstr "Feladó"
+msgstr "Feladó: %s"
 
 #: downloadmanager.py:292
 #, python-format
