From 560fddeebe7e0aa12a278bc7512bb0399b5f58da Mon Sep 17 00:00:00 2001
From: Milan Crha <mcrha@redhat.com>
Date: Fri, 6 Feb 2015 14:09:08 -0500
Subject: manager: port to latest EDS


diff --git a/configure.ac b/configure.ac
index 817b7e7..05dd33d 100644
--- a/configure.ac
+++ b/configure.ac
@@ -83,9 +83,9 @@ AC_SUBST(MAINTAINER_COMPILER_FLAGS)
 
 GLIB_MIN_VERSION=2.43.2
 GTK_MIN_VERSION=3.15.4
-ECAL_REQUIRED=3.5.3
-EDATASERVER_REQUIRED=3.5.3
-EDATASERVERUI_REQUIRED=3.5.3
+ECAL_REQUIRED=3.13.90
+EDATASERVER_REQUIRED=3.13.90
+EDATASERVERUI_REQUIRED=3.13.90
 LIBICAL_REQUIRED=0.43
 
 PKG_CHECK_MODULES(CALENDAR,
diff --git a/src/gcal-manager.c b/src/gcal-manager.c
index 3812d5f..68549fb 100644
--- a/src/gcal-manager.c
+++ b/src/gcal-manager.c
@@ -292,7 +292,7 @@ load_source (GcalManager *manager,
     {
       /* NULL: because maybe the operation cannot be really cancelled */
       e_cal_client_connect (source,
-                            E_CAL_CLIENT_SOURCE_TYPE_EVENTS, NULL,
+                            E_CAL_CLIENT_SOURCE_TYPE_EVENTS, 1, NULL,
                             on_client_connected,
                             manager);
     }
-- 
cgit v0.10.2

