From 6a859c65069df69f6a3f3e71624f6363fb276037 Mon Sep 17 00:00:00 2001
From: Alberto Milone <alberto.milone@canonical.com>
Date: Mon, 30 Jun 2014 12:16:26 +0200
Subject: [PATCH 1/1] Add support for Linux 3.16

---
 nv-linux.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/nv-linux.h b/nv-linux.h
index 9aac81a..075ee3f 100644
--- a/nv-linux.h
+++ b/nv-linux.h
@@ -210,6 +210,7 @@ extern int nv_pat_mode;
 #include <acpi/acpi_drivers.h>
 #if defined(NV_ACPI_DEVICE_OPS_HAS_MATCH) || defined(ACPI_VIDEO_HID)
 #define NV_LINUX_ACPI_EVENTS_SUPPORTED 1
+#include <acpi/acpi_bus.h>
 #endif
 #endif
 
-- 
1.9.1

