Index: ubuntuone-couch-0.3.0/ubuntuone/couch/auth.py
===================================================================
--- ubuntuone-couch-0.3.0.orig/ubuntuone/couch/auth.py	2011-12-27 15:49:54.000000000 -0500
+++ ubuntuone-couch-0.3.0/ubuntuone/couch/auth.py	2011-12-27 15:52:22.684217731 -0500
@@ -17,26 +17,29 @@
 import dbus
 import httplib2
 import logging
-import ubuntu_sso
 import urlparse
 
 from gettext import gettext as _
 from oauth import oauth
 from dbus.mainloop.glib import DBusGMainLoop
 
+from ubuntuone.platform.credentials.linux import (
+    DBUS_BUS_NAME,
+    DBUS_CREDENTIALS_PATH,
+)
+
 DBusGMainLoop(set_as_default=True)
 
 import socket
 socket.setdefaulttimeout(5)
 
-APP_NAME = "Ubuntu One"
 HMAC_SHA1 = oauth.OAuthSignatureMethod_HMAC_SHA1()
 PLAINTEXT = oauth.OAuthSignatureMethod_PLAINTEXT()
-NO_CREDENTIALS = _('No OAuth credentials passed in or found in Ubuntu SSO.')
+NO_CREDENTIALS = _('No OAuth credentials passed in or found in the system.')
 
 
 class CredentialsNotFound(Exception):
-    """Exception for missing data in SSO."""
+    """Exception for missing data in the credentials system."""
 
     def __init__(self, key):            # pylint: disable=W0231
         self.key = key
@@ -63,19 +66,22 @@
 
     DBusGMainLoop(set_as_default=True)
 
+    # Do not replicate this oauth code somewhere else, find_credentials_sync
+    # should NOT be used, use ubuntuone.credentials.CredentialsManagementTool
+    # instead.
+
     bus = dbus.SessionBus()
     proxy = bus.get_object(
-        ubuntu_sso.DBUS_BUS_NAME, ubuntu_sso.DBUS_CRED_PATH,
+        DBUS_BUS_NAME, DBUS_CREDENTIALS_PATH,
         follow_name_owner_changes=True)
     logging.info(
-        'get_oauth_data: asking for credentials from Ubuntu SSO. App name: %s',
-        APP_NAME)
+        'get_oauth_data: asking for credentials from the system.')
     oauth_data = dict(
         (_undbusify(k), _undbusify(v)) for k, v in
-        proxy.find_credentials(APP_NAME).iteritems())
+        proxy.find_credentials_sync().iteritems())
     if len(oauth_data) > 0:
         logging.info(
-            'get_oauth_data: Got non emtpy credentials from Ubuntu SSO.')
+            'get_oauth_data: Got non emtpy credentials from the system.')
     credentials = {}
     for key in ('consumer_key', 'consumer_secret', 'token', 'token_secret'):
         try:
@@ -119,8 +125,8 @@
     if access_token is None or consumer_key is None:
         try:
             credentials = get_oauth_credentials()
-        except CredentialsNotFound, e:
-            logging.error('No credentials found in SSO. %s', e)
+        except CredentialsNotFound:
+            logging.exception('No credentials found in the system:')
             return NO_CREDENTIALS
         access_token = credentials['token']
         token_secret = credentials['token_secret']
