diff -Naur Terminal-42.orig/terminal.py Terminal-42/terminal.py
--- Terminal-42.orig/terminal.py	2014-11-26 07:13:33.152002000 +0100
+++ Terminal-42/terminal.py	2014-11-26 07:14:02.372666787 +0100
@@ -18,7 +18,7 @@
 
 import os
 import sys
-import simplejson
+import json
 import ConfigParser
 import logging
 from gettext import gettext as _
@@ -389,7 +389,7 @@
 
         fd = open(file_path, 'r')
         text = fd.read()
-        data = simplejson.loads(text)
+        data = json.loads(text)
         fd.close()
 
         # Clean out any existing tabs.
@@ -438,7 +438,7 @@
             data['tabs'].append(tab_state)
 
         fd = open(file_path, 'w')
-        text = simplejson.dumps(data)
+        text = json.dumps(data)
         fd.write(text)
         fd.close()
 
