diff -Naur sugar-0.102.0.orig/extensions/cpsection/background/view.py sugar-0.102.0/extensions/cpsection/background/view.py
--- sugar-0.102.0.orig/extensions/cpsection/background/view.py	2014-04-22 15:53:58.000000000 +0200
+++ sugar-0.102.0/extensions/cpsection/background/view.py	2014-11-25 02:00:31.116882713 +0100
@@ -17,7 +17,7 @@
 
 import os
 
-import gi
+from gi.repository import GLib
 from gi.repository import Gtk
 from gi.repository import Gdk
 from gi.repository import GObject
@@ -122,7 +122,7 @@
                 pixbuf = GdkPixbuf.Pixbuf.new_from_file_at_size(
                     file_path, style.XLARGE_ICON_SIZE,
                     style.XLARGE_ICON_SIZE)
-            except gi._glib._glib.GError:
+            except GLib.GError:
                 pass
             else:
                 self._store.append([pixbuf, file_path])
