From 1980af084be40dc91a32817ce4c97cfd7248ebd1 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Gy=C3=B6rgy=20Ball=C3=B3?= <ballogyor@gmail.com>
Date: Mon, 13 Feb 2017 04:03:29 +0100
Subject: [PATCH] Don't crash if D-Bus method is not available

---
 src/sugar/presence/presenceservice.py | 8 ++------
 1 file changed, 2 insertions(+), 6 deletions(-)

diff --git a/src/sugar/presence/presenceservice.py b/src/sugar/presence/presenceservice.py
index b6c581e..7370710 100644
--- a/src/sugar/presence/presenceservice.py
+++ b/src/sugar/presence/presenceservice.py
@@ -88,12 +88,8 @@ def get_activity(self, activity_id, warn_if_none=True):
                             activity_id,
                             dbus_interface=CONN_INTERFACE_ACTIVITY_PROPERTIES)
                 except dbus.exceptions.DBusException, e:
-                    name = 'org.freedesktop.Telepathy.Error.NotAvailable'
-                    if e.get_dbus_name() == name:
-                        logging.debug("There's no shared activity with the id "
-                                      "%s", activity_id)
-                    else:
-                        raise
+                    logging.debug("There's no shared activity with the id "
+                                  "%s", activity_id)
                 else:
                     activity = Activity(account_path, connection.connection,
                                         room_handle=room_handle)
