diff -Naur Calculate-42.orig/toolbars.py Calculate-42/toolbars.py
--- Calculate-42.orig/toolbars.py	2014-11-28 06:28:14.185175000 +0100
+++ Calculate-42/toolbars.py	2014-11-28 06:30:45.876955401 +0100
@@ -9,7 +9,7 @@
 from mathlib import MathLib
 
 from sugar3.graphics.palette import Palette
-from sugar3.graphics.menuitem import MenuItem
+from sugar3.graphics.palettemenu import PaletteMenuBox, PaletteMenuItem
 from sugar3.graphics.toolbutton import ToolButton
 from sugar3.graphics.toggletoolbutton import ToggleToolButton
 from sugar3.graphics.style import GRID_CELL_SIZE
@@ -54,10 +54,12 @@
         p = Palette(text)
 
         if help_cb is not None:
-            item = MenuItem(_('Help'), 'action-help')
-            item.connect('activate', help_cb)
-            item.show()
-            p.menu.append(item)
+            menu_box = PaletteMenuBox()
+            p.set_content(menu_box)
+            menu_item = PaletteMenuItem(_('Help'), 'action-help')
+            menu_item.connect('activate', help_cb)
+            menu_box.append_item(menu_item)
+            menu_box.show_all()
 
         self.set_palette(p)
 
@@ -129,16 +131,13 @@
         copy_tool.connect('clicked', lambda x: calc.text_copy())
         self.insert(copy_tool, -1)
 
-        menu_item = MenuItem(_('Cut'))
-
-        try:
-            menu_item.set_accelerator(_('<ctrl>x'))
-        except AttributeError:
-            pass
-
+        menu_box = PaletteMenuBox()
+        copy_tool.get_palette().set_content(menu_box)
+        menu_item = PaletteMenuItem(_('Cut'))
+        menu_item.set_accelerator(_('<ctrl>x'))
         menu_item.connect('activate', lambda x: calc.text_cut())
-        menu_item.show()
-        copy_tool.get_palette().menu.append(menu_item)
+        menu_box.append_item(menu_item)
+        menu_box.show_all()
 
         self.insert(IconToolButton('edit-paste', _('Paste'),
             lambda x: calc.text_paste(),
